runtime/integration-tests/src/generic/envs/fudge_env.rs

```diff
@@ -1 +1,152 @@
-// TODO: implement generic::env::Env for an environment using fudge
+pub mod handle;
+
+use std::collections::HashMap;
+
+use cfg_primitives::{Balance, BlockNumber, Index};
+use fudge::primitives::Chain;
+use handle::{FudgeHandle, ParachainClient};
+use sc_client_api::HeaderBackend;
+use sp_api::{ApiRef, ProvideRuntimeApi};
+use sp_runtime::{generic::BlockId, DispatchError, DispatchResult, Storage};
+
+use crate::{
+	generic::{
+		environment::{utils, Env},
+		runtime::Runtime,
+	},
+	utils::accounts::Keyring,
+};
+
+/// Trait that represent a runtime with Fudge support
+pub trait FudgeSupport: Runtime {
+	/// Type to interact with fudge
+	type FudgeHandle: FudgeHandle<Self>;
+}
+
+/// Evironment that uses fudge to interact with the runtime
+pub struct FudgeEnv<T: Runtime + FudgeSupport> {
+	handle: T::FudgeHandle,
+	nonce_storage: HashMap<Keyring, Index>,
+}
+
+impl<T: Runtime + FudgeSupport> Env<T> for FudgeEnv<T> {
+	fn from_storage(storage: Storage) -> Self {
+		let mut handle = T::FudgeHandle::build(Storage::default(), storage);
+
+		handle.evolve();
+
+		Self {
+			handle,
+			nonce_storage: HashMap::default(),
+		}
+	}
+
+	fn submit_now(
+		&mut self,
+		_who: Keyring,
+		_call: impl Into<T::RuntimeCallExt>,
+	) -> Result<Balance, DispatchError> {
+		unimplemented!("FudgeEnv does not support submit_now() try submit_later()")
+	}
+
+	fn submit_later(&mut self, who: Keyring, call: impl Into<T::RuntimeCallExt>) -> DispatchResult {
+		let nonce = *self.nonce_storage.entry(who).or_default();
+
+		let extrinsic = self.state(|| utils::create_extrinsic::<T>(who, call, nonce));
+
+		self.handle
+			.parachain_mut()
+			.append_extrinsic(extrinsic)
+			.map(|_| ())
+			.map_err(|_| {
+				DispatchError::Other("Specific kind of DispatchError not supported by fudge now")
+				// More information, issue: https://github.com/centrifuge/fudge/issues/67
+			})?;
+
+		self.nonce_storage.insert(who, nonce + 1);
+
+		Ok(())
+	}
+
+	fn state_mut<R>(&mut self, f: impl FnOnce() -> R) -> R {
+		self.handle.parachain_mut().with_mut_state(f).unwrap()
```

Comment:
Should we maybe drop the `unwrap` and return the result here? Thinking that maybe the caller should then decide how to handle/unwrap/expect that further.

Reply:
Mmm.. do you know exactly which kind of events could `with_mut_state(f)` make fail?

From my point of view, if such events can not be handle by the tester before calling this call, then we should return the error, instead, if the tester could know before calling `with_mut_state()` which things they should change not to have an error, then an `unwrap()`  is ok because it acts as a precondition that any successful test must satisfy.

Reply:
Hmmm, only fudge-specific actions might trigger an error in this case, e.g. the backend failing to retrieve state at block X.

Reply:
And if it fails, it's an error from the tester making a wrong configuration for its test?

Reply:
Highly unlikely, maybe my initial suggestion can be disregarded for now - we could address this later if really needed.

Sorry for the noise.

Reply:
Absolutely not! Pretty good question to think about 👌🏻

Let's see then how it evolves keeping this in mind 👍🏻